backend/common/src/main/java/ai/verta/modeldb/common/db/migration/Migrator.java

```diff
@@ -136,40 +144,45 @@ private MigrationDatastore setupDatastore() throws SQLException {
     return migrationDatastore;
   }
 
-  private void runMigrations(
-      MigrationDatastore migrationDatastore, SortedSet<Migration> migrationsToPerform)
-      throws MigrationException, SQLException {
-    MigrationTools.lockDatabase(migrationDatastore);
-    try {
-      for (Migration migration : migrationsToPerform) {
-        try {
-          updateVersion(migration, true);
-          executeSingleMigration(migration);
-          updateVersion(migration, false);
-        } catch (IOException | SQLException e) {
-          throw new MigrationException("failed migration " + migration + "", e);
-        }
+  private void runMigrations(SortedSet<Migration> migrationsToPerform) throws MigrationException {
+    for (Migration migration : migrationsToPerform) {
+      try {
+        updateVersion(
+            migration, true, migration.isUp() ? migration.getNumber() - 1 : migration.getNumber());
+        executeSingleMigration(migration);
+        updateVersion(
+            migration, false, migration.isUp() ? migration.getNumber() : migration.getNumber() - 1);
+      } catch (IOException | SQLException e) {
+        throw new MigrationException("failed migration " + migration + "", e);
       }
-    } finally {
-      migrationDatastore.unlock();
     }
   }
 
-  private void updateVersion(Migration pendingMigration, boolean dirty) throws SQLException {
+  private void updateVersion(Migration pendingMigration, boolean dirty, int expectedCurrentVersion)
+      throws SQLException {
     int newVersion =
         pendingMigration.isUp() ? pendingMigration.getNumber() : pendingMigration.getNumber() - 1;
-    updateVersion(dirty, newVersion);
+
+    updateVersion(dirty, newVersion, expectedCurrentVersion);
   }
 
-  private void updateVersion(boolean dirty, int newVersion) throws SQLException {
+  private void updateVersion(boolean dirty, int newVersion, int expectedCurrentVersion)
```

Comment:
Maybe also check the dirty state as extra safety?

Reply:
good idea. will add!

Reply:
done